s3tests_boto3/functional/test_iam.py

```diff
@@ -1591,3 +1593,171 @@ def test_deny_version_id_condition_in_user_policy():
     response = s3_client_alt.delete_bucket(Bucket=bucket)
     eq(response['ResponseMetadata']['HTTPStatusCode'], 204)
 
+
+def test_copy_object_allow_put_dest_object():
+    client = get_iam_client()
+    # Create bucket user1buck as user1
+    s3_client_iam = get_iam_s3client()
+    user1buck = get_new_bucket(client=s3_client_iam)
+    # Create bucket user2buck and upload object srcobj as user2
+    s3_client_alt = get_alt_client()
+    user2buck = get_new_bucket(client=s3_client_alt)
+    response = s3_client_alt.put_object(Bucket=user2buck, Key='srcobj', Body='bar')
+    etag = response['ETag']
+    # Add inline user policy with Allow s3:PutObject on user1buck for user2
+    policy_document_allow = json.dumps(
+        {"Version": "2012-10-17",
+         "Statement": {
+             "Effect": "Allow",
+             "Action": ["s3:PutObject"],
+             "Resource": f"arn:aws:s3:::{user1buck}/*"}}
+    )
+    response = client.put_user_policy(PolicyDocument=policy_document_allow,
+                                      PolicyName='AllowPutPolicy', UserName=get_alt_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    # Copy object user2buck/srcobj to user1buck/dstobj  as user2
+    copy_source = {"Bucket": user2buck, 'Key': 'srcobj'}
+    response = s3_client_alt.copy_object(Bucket=user1buck, CopySource=copy_source, Key='dstobj')
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    response = client.delete_user_policy(PolicyName='AllowPutPolicy', UserName=get_alt_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    response = s3_client_alt.head_object(Bucket=user1buck, Key="dstobj")
```

Comment:
Should we add similar check for user2buck/srcobj also just to make sure after doing copy object original obj is not affected?

Reply:
Now reading Etag of original object and comparing.